File: app/src/main/java/net/feheren_fekete/applist/applistpage/ShortcutHelper.java
package net.feheren_fekete.applist.applistpage;

import android.annotation.TargetApi;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.pm.LauncherApps;
import android.content.pm.PackageManager;
import android.content.pm.ShortcutInfo;
import android.content.res.Resources;
import android.graphics.Bitmap;
import android.graphics.drawable.Drawable;
import android.os.Build;
import android.util.Log;
import android.widget.Toast;

import net.feheren_fekete.applist.ApplistLog;
import net.feheren_fekete.applist.R;
import net.feheren_fekete.applist.applistpage.model.AppShortcutData;
import net.feheren_fekete.applist.applistpage.model.ApplistModel;
import net.feheren_fekete.applist.applistpage.model.ShortcutData;
import net.feheren_fekete.applist.utils.ImageUtils;

import java.net.URISyntaxException;
import java.util.concurrent.Callable;

import bolts.Continuation;
import bolts.Task;

public class ShortcutHelper {

    private static final String TAG = ShortcutHelper.class.getSimpleName();
    private static final String ACTION_INSTALL_SHORTCUT = "com.android.launcher.action.INSTALL_SHORTCUT";

    // TODO: Inject
    private ApplistModel mApplistModel = ApplistModel.getInstance();

    private Context mContext;

    public ShortcutHelper(Context context) {
        mContext = context;
    }

    public void registerInstallShortcutReceiver() {
        IntentFilter intentFilter = new IntentFilter();
        intentFilter.addAction(ACTION_INSTALL_SHORTCUT);
        mContext.registerReceiver(mInstallShortcutReceiver, intentFilter);
    }

    public void unregisterInstallShortcutReceiver() {
        mContext.unregisterReceiver(mInstallShortcutReceiver);
    }

    public boolean handleIntent(Intent intent) {
        if (LauncherApps.ACTION_CONFIRM_PIN_SHORTCUT.equals(intent.getAction())) {
            handleShortcutRequest(intent);
            return true;
        }
        return false;
    }

    @TargetApi(Build.VERSION_CODES.O)
    private void handleShortcutRequest(Intent intent) {
        final LauncherApps launcherApps = (LauncherApps) mContext.getSystemService(Context.LAUNCHER_APPS_SERVICE);
        if (!launcherApps.hasShortcutHostPermission()) {
            return;
        }

        final LauncherApps.PinItemRequest pinItemRequest = intent.getParcelableExtra(LauncherApps.EXTRA_PIN_ITEM_REQUEST);
        final ShortcutInfo shortcutInfo = pinItemRequest.getShortcutInfo();
        Log.d(TAG, "PINNING " + shortcutInfo.getPackage() + " " + shortcutInfo.getId());

        if (!shortcutInfo.isEnabled()) {
            Toast.makeText(mContext, R.string.cannot_pin_disabled_shortcut, Toast.LENGTH_SHORT).show();
            return;
        }

        final String packageName = shortcutInfo.getPackage();
        final String shortcutId = shortcutInfo.getId();

        // BUG: Framework or app?
        //
        // We receive the LauncherApps.ACTION_CONFIRM_PIN_SHORTCUT intent twice.
        // First in MainActivity.onNewINtent() and then in MainActivity.onCreate(). Why is this?
        // Then we crash on the second call to LauncherApps.PinItemRequest.accept(). We are not allowed
        // to call it twice.
        //
        // Workaround: Use ApplistModel.hasInstalledAppShortcut() to check if we have already pinned
        // this shortcut (i.e. called accept() on it).
        //
        Task.callInBackground(new Callable<Boolean>() {
            @Override
            public Boolean call() throws Exception {
                return mApplistModel.hasInstalledAppShortcut(packageName, shortcutId);
            }
        }).continueWith(new Continuation<Boolean, Void>() {
            @Override
            public Void then(Task<Boolean> task) throws Exception {
                boolean isShortcutInstalled = task.getResult();
                if (isShortcutInstalled) {
                    Toast.makeText(mContext, R.string.cannot_pin_pinned_shortcut, Toast.LENGTH_SHORT).show();
                    return null;
                }
                if (!pinItemRequest.accept()) {
                    return null;
                }

                String shortcutName = null;
                if (shortcutInfo.getShortLabel() != null) {
                    shortcutName = shortcutInfo.getShortLabel().toString();
                } else if (shortcutInfo.getLongLabel() != null) {
                    shortcutName = shortcutInfo.getLongLabel().toString();
                } else {
                    ApplistLog.getInstance().log(new RuntimeException("Shortcut has no label"));
                    return null;
                }

                final Drawable iconDrawable = launcherApps.getShortcutBadgedIconDrawable(shortcutInfo, 0);
                final Bitmap shortcutIconBitmap = ImageUtils.drawableToBitmap(iconDrawable);

                final AppShortcutData shortcutData = new AppShortcutData(
                        System.currentTimeMillis(),
                        shortcutName,
                        packageName,
                        shortcutId);
                Task.callInBackground(new Callable<Void>() {
                    @Override
                    public Void call() throws Exception {
                        mApplistModel.addInstalledShortcut(shortcutData, shortcutIconBitmap);
                        return null;
                    }
                });
                return null;
            }
        }, Task.UI_THREAD_EXECUTOR);
    }

    private BroadcastReceiver mInstallShortcutReceiver = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            String action = intent.getAction();
            if (ACTION_INSTALL_SHORTCUT.equals(action)) {
                final String shortcutName = intent.getStringExtra(Intent.EXTRA_SHORTCUT_NAME);
                final Intent shortcutIntent = intent.getParcelableExtra(Intent.EXTRA_SHORTCUT_INTENT);
                Bitmap shortcutIconBitmap = intent.getParcelableExtra(Intent.EXTRA_SHORTCUT_ICON);
                if (shortcutIconBitmap == null) {
                    final Intent.ShortcutIconResource shortcutIconResource = intent.getParcelableExtra(Intent.EXTRA_SHORTCUT_ICON_RESOURCE);
                    if (shortcutIconResource != null) {
                        PackageManager packageManager = mContext.getPackageManager();
                        Resources resources = null;
                        try {
                            resources = packageManager.getResourcesForApplication(shortcutIconResource.packageName);
                            final int drawableId = resources.getIdentifier(shortcutIconResource.resourceName, null, null);
                            final Drawable drawable = resources.getDrawable(drawableId);
                            shortcutIconBitmap = ImageUtils.drawableToBitmap(drawable);
                        } catch (PackageManager.NameNotFoundException e) {
                            ApplistLog.getInstance().log(e);
                        }
                    }
                }
                if (shortcutIconBitmap == null) {
                    ApplistLog.getInstance().log(new RuntimeException("Missing icon for shortcut: " + shortcutIntent.toUri(0)));
                    return;
                }

                final ShortcutData shortcutData = new ShortcutData(
                        System.currentTimeMillis(),
                        shortcutName,
                        shortcutIntent);
                final Bitmap finalShortcutIconBitmap = shortcutIconBitmap;
                Task.callInBackground(new Callable<Void>() {
                    @Override
                    public Void call() throws Exception {
                        mApplistModel.addInstalledShortcut(shortcutData, finalShortcutIconBitmap);
                        return null;
                    }
                });
            }
        }
    };

}

File: app/src/main/java/net/feheren_fekete/applist/applistpage/ShortcutHelper.kt
package net.feheren_fekete.applist.applistpage

import android.annotation.TargetApi
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.content.pm.LauncherApps
import android.content.pm.PackageManager
import android.content.pm.ShortcutInfo
import android.content.res.Resources
import android.graphics.Bitmap
import android.graphics.drawable.Drawable
import android.os.Build
import android.util.Log
import android.widget.Toast
import kotlinx.coroutines.*
import net.feheren_fekete.applist.ApplistLog
import net.feheren_fekete.applist.R
import net.feheren_fekete.applist.applistpage.repository.ApplistPageRepository
import net.feheren_fekete.applist.applistpage.repository.database.ApplistItemData
import net.feheren_fekete.applist.utils.ImageUtils
import org.koin.java.KoinJavaComponent.get
import org.koin.java.KoinJavaComponent.inject


class ShortcutHelper {

    // This must be get() to avoid crash due to lazy-init (inject())
    // inside coroutine threads.
    // Crash is caused by creating a Handler in ApplistModel. Inside
    // threads we cannot create handles by "new Handle()".
    private val applistRepo = get(ApplistPageRepository::class.java)
    private val applistLog = get(ApplistLog::class.java)

    private val imageUtils: ImageUtils by inject(ImageUtils::class.java)

    private val installShortcutReceiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            val action = intent.action
            if (ACTION_INSTALL_SHORTCUT == action) {
                applistLog.analytics(ApplistLog.CREATE_LEGACY_SHORTCUT, ApplistLog.OTHER_APP)
                val shortcutName: String? = intent.getStringExtra(Intent.EXTRA_SHORTCUT_NAME)
                if (shortcutName == null) {
                    applistLog.log(RuntimeException("Missing shortcut name: ${intent.toUri(0)}"))
                    return
                }
                val shortcutIntent: Intent? = intent.getParcelableExtra<Intent>(Intent.EXTRA_SHORTCUT_INTENT)
                if (shortcutIntent == null) {
                    applistLog.log(RuntimeException("Missing shortcut intent: ${intent.toUri(0)}"))
                    return
                }
                var shortcutIconBitmap: Bitmap? = intent.getParcelableExtra(Intent.EXTRA_SHORTCUT_ICON)
                if (shortcutIconBitmap == null) {
                    val shortcutIconResource = intent.getParcelableExtra<Intent.ShortcutIconResource>(Intent.EXTRA_SHORTCUT_ICON_RESOURCE)
                    if (shortcutIconResource != null) {
                        val packageManager = context.packageManager
                        try {
                            val resources = packageManager.getResourcesForApplication(shortcutIconResource.packageName)
                            val drawableId = resources!!.getIdentifier(shortcutIconResource.resourceName, null, null)
                            val drawable = resources.getDrawable(drawableId, null)
                            shortcutIconBitmap = imageUtils.drawableToBitmap(drawable)
                        } catch (e: Resources.NotFoundException) {
                            applistLog.log(e)
                        } catch (e: PackageManager.NameNotFoundException) {
                            applistLog.log(e)
                        } catch (e: NullPointerException) {
                            applistLog.log(e)
                        }
                    }
                }
                if (shortcutIconBitmap == null) {
                    applistLog.log(RuntimeException("Missing icon for shortcut: " + shortcutIntent.toUri(0)))
                    shortcutIconBitmap = imageUtils.shortcutPlaceholder()
                }

                val packageName = shortcutIntent.getPackage()
                        ?: shortcutIntent.component?.packageName
                if (packageName == null) {
                    applistLog.log(RuntimeException("Missing package for shortcut: " + shortcutIntent.toUri(0)))
                    return
                }

                val shortcutData = ApplistItemData.createShortcut(
                        0,
                        packageName,
                        shortcutName,
                        "",
                        shortcutIntent)
                GlobalScope.launch(Dispatchers.IO) {
                    applistRepo.addShortcut(
                            shortcutData, shortcutIconBitmap!!,
                            ApplistItemData.DEFAULT_SECTION_ID, false)
                }
            }
        }
    }

    fun registerInstallShortcutReceiver(context: Context) {
        val intentFilter = IntentFilter()
        intentFilter.addAction(ACTION_INSTALL_SHORTCUT)
        context.registerReceiver(installShortcutReceiver, intentFilter)
    }

    fun unregisterInstallShortcutReceiver(context: Context) {
        context.unregisterReceiver(installShortcutReceiver)
    }

    fun handleIntent(context: Context, intent: Intent): Boolean {
        val action = intent.action
        if (LauncherApps.ACTION_CONFIRM_PIN_SHORTCUT == action) {
            applistLog.analytics(ApplistLog.CREATE_PINNED_APP_SHORTCUT, ApplistLog.OTHER_APP)
            handleShortcutRequest(context, intent)
            return true
        }
        return false
    }

    @TargetApi(Build.VERSION_CODES.O)
    fun pinAppShortcut(coroutineScope: CoroutineScope,
                       context: Context,
                       shortcutInfo: ShortcutInfo,
                       pinItemRequest: LauncherApps.PinItemRequest?) {
        val launcherApps = context.getSystemService(Context.LAUNCHER_APPS_SERVICE) as LauncherApps
        if (!launcherApps.hasShortcutHostPermission()) {
            return
        }

        Log.d(TAG, "PINNING " + shortcutInfo.getPackage() + " " + shortcutInfo.id)

        if (!shortcutInfo.isEnabled) {
            Toast.makeText(context, R.string.cannot_pin_disabled_shortcut, Toast.LENGTH_SHORT).show()
            return
        }

        val packageName = shortcutInfo.getPackage()
        val shortcutId = shortcutInfo.id

        // BUG: Framework or app?
        //
        // We receive the LauncherApps.ACTION_CONFIRM_PIN_SHORTCUT intent twice.
        // First in MainActivity.onNewINtent() and then in MainActivity.onCreate(). Why is this?
        // Then we crash on the second call to LauncherApps.PinItemRequest.accept(). We are not allowed
        // to call it twice.
        //
        // Workaround: Use ApplistModel.hasAppShortcut() to check if we have already pinned
        // this shortcut (i.e. called accept() on it).
        //
        coroutineScope.launch(Dispatchers.Main) {
            val isShortcutInstalled = withContext(Dispatchers.IO) {
                applistRepo.hasAppShortcut(packageName, shortcutId)
            }

            if (isShortcutInstalled) {
                Toast.makeText(context, R.string.cannot_pin_pinned_shortcut, Toast.LENGTH_SHORT).show()
                return@launch
            }

            if (pinItemRequest != null) {
                try {
                    if (!pinItemRequest.accept()) {
                        return@launch
                    }
                } catch (e: Exception) {
                    Toast.makeText(context, R.string.cannot_pin, Toast.LENGTH_SHORT).show()
                    applistLog.log(e)
                    return@launch
                }
            }

            val shortcutName = shortcutInfo.shortLabel?.toString()
                    ?: shortcutInfo.longLabel?.toString()
            if (shortcutName == null) {
                Toast.makeText(context, R.string.cannot_pin, Toast.LENGTH_SHORT).show()
                applistLog.log(RuntimeException("Shortcut has no label"))
                return@launch
            }

            val iconDrawable: Drawable? = try {
                launcherApps.getShortcutBadgedIconDrawable(shortcutInfo, 0)
            } catch (e: Exception) {
                applistLog.log(RuntimeException("LauncherApps.getShortcutBadgedIconDrawable"))
                try {
                    launcherApps.getShortcutIconDrawable(shortcutInfo, 0)
                } catch (e: Exception) {
                    applistLog.log(RuntimeException("LauncherApps.getShortcutIconDrawable"))
                    null
                }
            }
            val shortcutIconBitmap = if (iconDrawable != null) {
                imageUtils.drawableToBitmap(iconDrawable)
            } else {
                imageUtils.shortcutPlaceholder()
            }

            val shortcutData = ApplistItemData.createAppShortcut(
                    0,
                    shortcutName,
                    "",
                    packageName,
                    shortcutId)
            launch(Dispatchers.IO) {
                applistRepo.addShortcut(
                        shortcutData, shortcutIconBitmap,
                        ApplistItemData.DEFAULT_SECTION_ID, false)
            }
        }
    }

    @TargetApi(Build.VERSION_CODES.O)
    private fun handleShortcutRequest(context: Context, intent: Intent) {
        val pinItemRequest = intent.getParcelableExtra<LauncherApps.PinItemRequest>(LauncherApps.EXTRA_PIN_ITEM_REQUEST)
        val shortcutInfo = pinItemRequest.shortcutInfo
        if (shortcutInfo == null) {
            return
        }
        pinAppShortcut(GlobalScope, context, shortcutInfo, pinItemRequest)
    }

    companion object {
        private val TAG = ShortcutHelper::class.java.simpleName
        private const val ACTION_INSTALL_SHORTCUT = "com.android.launcher.action.INSTALL_SHORTCUT"
    }

}
